mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java

```diff
@@ -28,28 +29,62 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class SynchronousEventDelivery implements EventDelivery {
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(SynchronousEventDelivery.class);
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class EventDeliveryImpl implements EventDelivery {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventDeliveryImpl.class);
 
     private final MetricFactory metricFactory;
 
-    @Inject
-    public SynchronousEventDelivery(MetricFactory metricFactory) {
+    public EventDeliveryImpl(MetricFactory metricFactory) {
         this.metricFactory = metricFactory;
     }
 
     @Override
-    public void deliver(MailboxListener mailboxListener, Event event) {
+    public ExecutionStages deliver(Collection<MailboxListener> mailboxListeners, Event event) {
+        CompletableFuture<Void> synchronousListeners = doDeliver(
+            filterByExecutionMode(mailboxListeners, MailboxListener.ExecutionMode.SYNCHRONOUS),
+            event);
+        CompletableFuture<Void> asyncListener = doDeliver(
+            filterByExecutionMode(mailboxListeners, MailboxListener.ExecutionMode.ASYNCHRONOUS),
+            event);
+
+        return new ExecutionStages(synchronousListeners, asyncListener);
+    }
+
+    private ImmutableList<MailboxListener> filterByExecutionMode(Collection<MailboxListener> mailboxListeners, MailboxListener.ExecutionMode executionMode) {
+        return mailboxListeners.stream()
+            .filter(listener -> listener.getExecutionMode() == executionMode)
+            .collect(Guavate.toImmutableList());
+    }
+
+    private CompletableFuture<Void> doDeliver(Collection<MailboxListener> mailboxListeners, Event event) {
+        return Flux.fromIterable(mailboxListeners)
+            .flatMap(mailboxListener -> Mono.fromCallable(() ->  {
```

Comment:
what prevents you from using ParallelFlux ?

Reply:
I thought flatMap was doing this for me?

Reply:
In theory, yes. I don't know exactly if there's a difference. Let's keep your code.